### ISSUE TITLE
`redundantSelf` rule breaks build in case with complex switch statement

### ISSUE THREAD

**Issue by calda** [nicklockwood/SwiftFormat#1304]
The `redundantSelf` rule breaks this sample code:

```swift
class Foo {
    let bar = true
    let someOptionalBar: String? = "bar"

    func test() {
        guard let bar: String = someOptionalBar else {
            return
        }

        let result = Result<Any, Error>.success(bar)
        switch result {
        case let .success(value):
            if self.bar {
                if self.bar {
                    print(self.bar)
                }
            } else {
                if self.bar {
                    print(self.bar)
                }
            }
        case .failure:
            if self.bar {
                print(self.bar)
            }
        }
    }
}
```

by converting it to:

```swift
class Foo {
    let bar = true
    let someOptionalBar: String? = "bar"

    func test() {
        guard let bar: String = someOptionalBar else {
            return
        }

        let result = Result<Any, Error>.success(bar)
        switch result {
        case let .success(value):
            if self.bar {
                if self.bar {
                    print(self.bar)
                }
            } else {
                if bar {        // <---- self. was removed
                    print(bar)  // <---- self. was removed
                }
            }
        case .failure:
            if bar {        // <---- self. was removed
                print(bar)  // <---- self. was removed
            }
        }
    }
}
```

which no longer compiles

**Comment by calda** [comment]
Here's another case that fails for the same reason:

```swift
func testClosureParameterListShadowingPropertyOnSelf() {
    let input = """
    class Foo {
        func method() {
            let result: Result<Any?, Error> = .success("foo")
            switch result {
            case let .success(value):
                if let value {
                    print(value)
                }
                self.method()
            case .failure:
                break
            }
        }
    }
    """

    testFormatting(for: input, rule: FormatRules.redundantSelf)
}
```

**Comment by calda** [comment]
Here's a case that demonstrates the same issue but as it relates to #1303:

```swift
func testSelfCallAfterIfStatementInSwitchStatement() {
    let input = """
    closure { [weak self] in
        guard let self else {
            return
        }

        switch result {
        case let .success(value):
            if value != nil {
                if value != nil {
                    self.method()
                }
            }
            self.method() // incorrectly updated to `method()`
        case .failure:
            break
        }
    }
    """

    let options = FormatOptions(swiftVersion: "5.3")
    testFormatting(for: input, rule: FormatRules.redundantSelf, options: options)
}
```

**Comment by nicklockwood** [comment]
Seems to have been the same root cause as https://github.com/nicklockwood/SwiftFormat/issues/1305 - both are now fixed

**Comment by calda** [comment]
Thanks!

**Comment by calda** [comment]
@nicklockwood, here's another case that still has the same issue:

```swift
func testSelfNotRemovedFollowingNestedSwitchStatements() {
    let input = """
    class Foo {
        let bar = true
        let someOptionalBar: String? = "bar"

        func test() {
            guard let bar: String = someOptionalBar else {
                return
            }

            let result = Result<Any, Error>.success(bar)
            switch result {
            case let .success(value):
                switch result {
                case .success:
                    print("success")
                case .value:
                    print("value")
                }
            case .failure:
                guard self.bar else {
                    print(self.bar)   // incorrectly updated to print(bar)
                    return
                }
                print(self.bar)  // incorrectly updated to print(bar)
            }
        }
    }
    """

    testFormatting(for: input, rule: FormatRules.redundantSelf)
}
```

This is less widespread than the other issue, I only see it once in our codebase (so it's easy to work around with `// swiftformat:disable:next redundantSelf`).

**Comment by nicklockwood** [comment]
@calda I found the problem - will post a fix later.

**Comment by nicklockwood** [comment]
Fixed

**Comment by calda** [comment]
Thank you! Always appreciate it

**Comment by nicklockwood** [comment]
@calda fixed in 0.50.4
